component/notifications.libsonnet

```
@@ -0,0 +1,256 @@
+      location: std.get(args, 'location', 'BannerTop'),
+      color: std.get(args, 'color', '#fff'),
+      backgroundColor: std.get(args, 'backgroundColor', '#2596be'),
+      link: std.get(args, 'link'),
+    },
+  };
+
+local consoleNotifications = [
+  makeConsoleNotification(name, params.notifications[name])
+  for name in std.objectFields(params.notifications)
+  if params.notifications[name] != null
+];
+
+local nextChannelOverlay() =
+  local overlays = inv.parameters.openshift_upgrade_controller.cluster_version.overlays;
+  local channelOverlays = {
+    [date]: overlays[date].spec.channel
+    for date in std.objectFields(overlays)
+    if std.objectHas(overlays[date].spec, 'channel')
+       && std.split(overlays[date].spec.channel, '.')[1] > params.openshift_version.Minor
+  };
+  local date = if std.length(channelOverlays) > 0 then
+    std.sort(std.objectFields(channelOverlays))[0];
+  if date != null then
+    {
+      date: date,
+      channel: channelOverlays[date],
+      version: std.split(channelOverlays[date], '-')[1],
+    };
+
+local upgradeControllerNS = {
+  metadata+: {
+    namespace: inv.parameters.openshift_upgrade_controller.namespace,
+  },
+};
+
+local notificationRBAC =
+  local argocd_sa = kube.ServiceAccount('notification-manager') + namespace;
+  local upgrade_sa = argocd_sa + upgradeControllerNS;
+  local cluster_role = kube.ClusterRole('appuio:upgrade-notification-editor') {
+    rules: [
+      {
+        apiGroups: [ 'console.openshift.io' ],
+        resources: [ 'consolenotifications' ],
+        verbs: [ '*' ],
+      },
+      {
+        apiGroups: [ 'managedupgrade.appuio.io' ],
+        resources: [ 'upgradeconfigs' ],
+        verbs: [ 'get', 'list' ],
+      },
+      // needed so that `oc version` can get the OCP server version
+      {
+        apiGroups: [ 'config.openshift.io' ],
+        resources: [ 'clusterversions' ],
+        verbs: [ 'get', 'list' ],
+      },
+      {
+        apiGroups: [ '' ],
+        resources: [ 'configmaps' ],
+        resourceNames: [ 'upgrade-notification-template' ],
+        verbs: [ '*' ],
+      },
+    ],
+  };
+  local cluster_role_binding =
+    kube.ClusterRoleBinding('appuio:upgrade-notification-manager') {
+      subjects_: [ argocd_sa, upgrade_sa ],
+      roleRef_: cluster_role,
+    };
+  {
+    argocd_sa: argocd_sa,
+    upgrade_sa: upgrade_sa,
+    cluster_role: cluster_role,
+    cluster_role_binding: cluster_role_binding,
+  };
+
+local createUpgradeNotification(overlay) =
+  [
+    kube.ConfigMap('upgrade-notification-template') + namespace {
+      data: {
+        'upgrade.yaml': std.manifestYamlDoc(
+          std.mergePatch(
+            makeConsoleNotification('upgrade-%s' % overlay.version, params.upgrade_notification.notification),
+            {
+              metadata: {
+                labels: {
+                  'appuio.io/ocp-version': overlay.version,
+                },
+              },
+            },
+          ),
```

The `std.mergePatch()` seems unnecessary here, unless I'm missing something. The following should be sufficient:

```suggestion
          makeConsoleNotification('upgrade-%s' % overlay.version, params.upgrade_notification.notification) {
            metadata+: {
              labels+: {
                'appuio.io/ocp-version': overlay.version,
              },
            },
          },
```

---

I thought so too, but somehow couldn't make it work, then I went for `std.mergePatch`. I'll give it another try